liblwgeom/lwgeom_median.c

```diff
@@ -115,37 +119,58 @@ iterate_3d(POINT3D* curr, const POINT3D* points, uint32_t npoints, double* dista
 	return delta;
 }
 
-static POINT3D
-init_guess(const POINT3D* points, uint32_t npoints)
+static POINT4D
+init_guess(const POINT4D* points, size_t npoints)
 {
-	POINT3D guess = { 0, 0, 0 };
-	uint32_t i;
+	POINT4D guess = { 0, 0, 0, 0 };
+	size_t i;
 	for (i = 0; i < npoints; i++)
 	{
-		guess.x += points[i].x / npoints;
-		guess.y += points[i].y / npoints;
-		guess.z += points[i].z / npoints;
+		guess.x += points[i].x * points[i].m;
+		guess.y += points[i].y * points[i].m;
+		guess.z += points[i].z * points[i].m;
+		guess.m += points[i].m;
+	}
+	if (!FP_IS_ZERO(guess.m))
+	{
+		guess.x /= guess.m;
+		guess.y /= guess.m;
+		guess.z /= guess.m;
 	}
-
 	return guess;
 }
 
-static POINT3D*
-lwmpoint_extract_points_3d(const LWMPOINT* g, uint32_t* ngeoms)
+static POINT4D*
+lwmpoint_extract_points_4d(const LWMPOINT* g, size_t* ngeoms)
 {
-	uint32_t i;
-	uint32_t n = 0;
+	size_t i;
+	size_t n = 0;
 	int is_3d = lwgeom_has_z((LWGEOM*) g);
+	int has_m = lwgeom_has_m((LWGEOM*) g);
 
-	POINT3D* points = lwalloc(g->ngeoms * sizeof(POINT3D));
+	POINT4D* points = lwalloc(g->ngeoms * sizeof(POINT4D));
 	for (i = 0; i < g->ngeoms; i++)
 	{
 		LWGEOM* subg = lwcollection_getsubgeom((LWCOLLECTION*) g, i);
 		if (!lwgeom_is_empty(subg))
 		{
-			getPoint3dz_p(((LWPOINT*) subg)->point, 0, (POINT3DZ*) &points[n++]);
+			getPoint4d_p(((LWPOINT*) subg)->point, 0, (POINT4D*) &points[n]);
 			if (!is_3d)
-				points[n-1].z = 0.0; /* in case the getPoint functions return NaN in the future for 2d */
+			{
+				points[n].z = 0.0; /* in case the getPoint functions return NaN in the future for 2d */
+			}
+			if (!has_m)
+			{
+				points[n].m = 1.0;
+				n++;
+			}
+			else
+			{
+				if (!FP_IS_ZERO(points[n].m))
+				{
+					n++;
```

Comment:
Add comment that points with zero weight are ignored

Reply:
What happens when weights are negative?

Reply:
A point with negative weight will cancel out a point with positive weight. When total sum of weights is negative, the point will go to infinity. Need to add a check for that.